kernel/rtlil.cc

```diff
@@ -200,17 +201,30 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
 	return res;
 }
 
-RTLIL::Const::Const(const std::string &str)
-{
-	flags = RTLIL::CONST_FLAG_STRING;
-	bits.reserve(str.size() * 8);
-	for (int i = str.size()-1; i >= 0; i--) {
-		unsigned char ch = str[i];
-		for (int j = 0; j < 8; j++) {
-			bits.push_back((ch & 1) != 0 ? State::S1 : State::S0);
-			ch = ch >> 1;
-		}
-	}
+PRIVATE_NAMESPACE_BEGIN
+
+using CharToBitsLUT = std::array<std::array<RTLIL::State, 8>, 256>;
+
+const CharToBitsLUT build_CharToBitsLUT() {
+        CharToBitsLUT LUT;
+        for (int i = 0; i < 256; ++i) {
+          unsigned char ch = i;
+          for (int j = 0; j < 8; j++) {
+            LUT[i][j] = (ch & 1) != 0 ? State::S1 : State::S0;
+            ch = ch >> 1;
+          }
+        }
+        return LUT;
+}
+PRIVATE_NAMESPACE_END
+
+RTLIL::Const::Const(const std::string &str) : flags(RTLIL::CONST_FLAG_STRING) {
+        static const CharToBitsLUT LUT = build_CharToBitsLUT();
+        bits.reserve(str.size() * 8);
+        for (int i = str.size() - 1; i >= 0; --i) {
+          const unsigned char ch = str[i];
+          bits.insert(bits.end(), LUT[ch].begin(), LUT[ch].end());
```

Comment:
I am bit surprised by this being faster -- would you perhaps have a figure for this change alone?

Reply:
This in itself is about 20% faster.  Decode_string is about 30% faster.

Reply:
Before:
![image](https://github.com/YosysHQ/yosys/assets/16907534/4f23cb5e-dec2-4f94-ba1c-68f3f7fda43c)

After:
![image](https://github.com/YosysHQ/yosys/assets/16907534/2ebc2de5-6b22-45e4-9252-03970a65db2f)